source/common/http/conn_manager_impl.cc

```diff
@@ -1460,14 +1464,32 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ResponseHeaderMap& heade
     }
   }
 
-  chargeStats(headers);
-
   ENVOY_STREAM_LOG(debug, "encoding headers via codec (end_stream={}):\n{}", *this, end_stream,
                    headers);
 
   // Now actually encode via the codec.
   filter_manager_.streamInfo().onFirstDownstreamTxByteSent();
-  response_encoder_->encodeHeaders(headers, end_stream);
+  const auto status = response_encoder_->encodeHeaders(headers, end_stream);
+  if (!status.ok()) {
+    // This branch can happen when a misbehaving filter chain removed critical headers or set
+    // malformed header values.
+
+    // We reenter this function after the local reply is sent below and then
+    // we again call `streamInfo().onFirstDownstreamTxByteSent()` above, which results in
+    // ASSERT(first_downstream_tx_byte_sent_ != absl::nullopt) failure.
+    // That's why we need to reset first_downstream_tx_byte_sent_ here.
+    // The byte has not actually been sent at this point.
+    filter_manager_.streamInfo().first_downstream_tx_byte_sent_ = absl::nullopt;
```

Comment:
This may be a question for senior maintainers but could `filter_manager_.streamInfo().onFirstDownstreamTxByteSent();` be called after `encodeHeaders` in a branch `if (status.ok())`? I don't know if that will mess up something up re timing (expecting time to be logged before or after).

Reply:
done: https://github.com/envoyproxy/envoy/pull/15658/commits/e4d8e9f8a3b07cfd9aec1541cb6721eb35b3ec16
at least all the tests pass when call it after the encoding